Sources/Data/Attributes.swift

```diff
@@ -0,0 +1,299 @@
+//
```

Comment:
In general, I found the code difficult to reason about due to the mixing of inline code logic and private extensions. I've added review suggestions throughout the code. Here's a summary of how this could be simplified:
 
 1. Initializer accepts a [String: Any] and validates all keys/values comply with our restrictions. Mark this initializer with `throws` instead of using `assertionFailure` so we can move the error handling strategy upstream.
 2. The NSCoding encode function wraps the already validated rawValue property into an NSDictionary and encodes it.
 3. The NSCoding decode function decodes an NSDictionary, converts it to [String: Any] and calls the initializer mentioned above.
 4. The Codable encode function encodes the already validated rawValue property into by using a casting pattern match in a switch statement.
 5. The Codable decode function decodes a [String: Any] type by attempting to decode each of the known valid types, then finally calls the initializer mentioned above to guarantee the `rawValue` property is valid.